08openmp/index.md

```diff
@@ -0,0 +1,307 @@
+
+In order to perform truly parallel computation, that is two completely different operations happening at the exact same moment, CPUs had to become **multi-core**, hosting many processing units within one CPU and using them to run many threads in parallel. In consumer PCs this transition started with the dual-core Intel Core2Duo series in the mid-2000s, but supercomputers were already playing with the slightly related idea of **multi-CPU** systems with hundreds of CPUs as early as the 80s! As of 2023, core counts on both server and consumer CPUs can be as high as 64. Typical consumer systems will have only one CPU while servers may have several on one motherboard. Writing parallel programs for multi-core machines is made easy using the focus of this week: **OpenMP**.
+
+Up until the 90s, supercomputers were dominated by **shared-memory** systems, where all processing units accessed one single memory space. In the late-80s/early-90s it become economically more feasible to increase parallelism and computational power by creating **distributed** systems made up of multiple **nodes**. Each node is a single computer with its own CPU(s), memory space, networking hardware, and possibly other components like hard drives, all networked together in such a way they can communicate information and operate as *one single parallel machine*. I'll start using the term **node** to refer to a single computer as part of a larger, distributed parallel machine. While shared-memory systems still exist[^jq-terabytes], nearly all the world's largest computers are now distributed systems.
+
+Distributed systems tend to have separate memory spaces for each node[^multi-cpu] which introduces a new complexity: how do the nodes share data or communicate when necessary? These **distributed-memory** systems are commonly programmed using a library called the **Message Passing Interface** (MPI) which allows individual nodes to communicate by passing data back and forth. We'll get into MPI properly in the last part of this course. For now, it's useful just to understand the difference between the two basic forms of parallel computing: **shared-memory** parallelism where each thread shares the same memory space and can access any piece of data within it, and **distributed-memory** systems where threads do not share one memory space and must explicitly send and receive data when necessary.
```

Comment:
Technicality, but MPI is the standard rather than a library. The actual library could be OpenMPI, Intel MPI, etc.